neonvm/controllers/alerttracker/tracker.go

```diff
@@ -0,0 +1,87 @@
+package alerttracker
+
+import "time"
+
+var Now = time.Now
+
+// Tracker is a struct that keeps track of failure/success events for a given key.
+// Failure state over a given interval will trigger a firing alert.
+type Tracker[T comparable] struct {
+	interval time.Duration
+
+	pendingSince map[T]time.Time
+	firing       map[T]struct{}
+	fireAt       []fireAt[T]
+}
+
+type fireAt[T comparable] struct {
+	ts  time.Time
+	key T
+}
+
+func NewTracker[T comparable](interval time.Duration) *Tracker[T] {
+	return &Tracker[T]{
+		interval:     interval,
+		pendingSince: make(map[T]time.Time),
+		firing:       make(map[T]struct{}),
+		fireAt:       []fireAt[T]{},
+	}
+}
+
+func (t *Tracker[T]) forward(now time.Time) {
```

Comment:
The logic here seems tricky... it'd be good to:

1. Have a brief high-level explainer about what's going on
2. Have some tests for multiple keys